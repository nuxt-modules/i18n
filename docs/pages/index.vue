<script setup lang="ts">
import { useClipboard } from '@vueuse/core'
const { data: page } = await useAsyncData('index', () => queryContent('/').findOne())

// Page Metadata (SEO & OG)
const { setPageMeta } = usePageMeta()
setPageMeta({
  title: page.value.title,
  description: page.value.description,
  headline: page.value.hero.headline?.label
})
const source = ref('npx nuxi module add i18n')
const { copy, copied } = useClipboard({ source })
</script>

<template>
  <div>
    <ULandingHero
      v-if="page.hero"
      v-bind="page.hero"
      :ui="{
        container: 'overflow-hidden py-10 flex flex-row items-center  gap-1',
        links: 'flex items-center gap-2',
        description: 'text-gray-500 dark:text-gray-400 text-xl max-w-2xl leading-normal mb-10'
      }"
    >
      <template #headline>
        <UBadge v-if="page.hero.headline" variant="subtle" size="lg" class="relative rounded-full font-semibold">
          <NuxtLink :to="page.hero.headline.to" target="_blank" class="focus:outline-none" tabindex="-1">
            <span class="absolute inset-0" aria-hidden="true" />
          </NuxtLink>

          {{ page.hero.headline?.label }}

          <UIcon
            v-if="page.hero.headline.icon"
            :name="page.hero.headline.icon"
            class="ml-1 w-4 h-4 pointer-events-none"
          />
        </UBadge>
      </template>

      <template #description>
        {{ page.hero.description }}
      </template>

      <template #title>
        <p>
          <span class="md:hidden">i18n</span>
          <span class="hidden md:block">Internationalization</span>
          for
          <span class="text-primary">Nuxt Applications</span>
        </p>
      </template>

      <UInput
        aria-label="Copy code to get started"
        :model-value="source"
        name="get-started"
        class="mx-auto"
        disabled
        autocomplete="off"
        size="lg"
        :ui="{ base: 'w-[300px] disabled:cursor-default', icon: { trailing: { pointer: '' } } }"
      >
        <template #leading>
          <UIcon name="i-ph-terminal" />
        </template>
        <template #trailing>
          <UButton
            aria-label="Copy Code"
            :color="copied ? 'green' : 'gray'"
            variant="ghost"
            :padded="false"
            :icon="copied ? 'i-ph-check' : 'i-ph-copy'"
            @click="copy(source)"
          />
        </template>
      </UInput>
    </ULandingHero>

    <ULandingSection :title="page.features.title">
      <UPageGrid>
        <ULandingCard v-for="(item, index) of page.features.items" :key="index" v-bind="item" />
      </UPageGrid>
    </ULandingSection>
  </div>
</template>
