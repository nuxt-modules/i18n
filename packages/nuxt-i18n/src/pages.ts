import createDebug from 'debug'
import { extendPages } from '@nuxt/kit'
import { localizeRoutes } from 'vue-i18n-routing'

import type { Nuxt, NuxtPage } from '@nuxt/schema'
import type { RouteOptionsResolver, ComputedRouteOptions } from 'vue-i18n-routing'
import type { NuxtI18nOptions, CustomRoutePages } from './types'

const debug = createDebug('@nuxtjs/i18n:pages')

export function setupPages(
  options: Required<NuxtI18nOptions>,
  nuxt: Nuxt,
  additionalOptions: { isBridge?: boolean; localeCodes: string[] } = {
    isBridge: false,
    localeCodes: []
  }
) {
  let includeUprefixedFallback = nuxt.options.target === 'static'
  nuxt.hook('generate:before', () => {
    debug('called generate:before hook')
    includeUprefixedFallback = true
  })

  const pagesDir = nuxt.options.dir && nuxt.options.dir.pages ? nuxt.options.dir.pages : 'pages'
  const { trailingSlash } = nuxt.options.router
  debug(`pagesDir: ${pagesDir}, tailingSlash: ${trailingSlash}`)

  extendPages(pages => {
    debug('pages making ...', pages)
    const localizedPages = localizeRoutes(pages, {
      ...options,
      includeUprefixedFallback,
      optionsResolver: getRouteOptionsResolver(options.pages, pagesDir, options.defaultLocale)
    })
    pages.splice(0, pages.length)
    pages.unshift(...(localizedPages as NuxtPage[]))
    debug('... made pages', pages)
  })
}

function getRouteOptionsResolver(
  pages: CustomRoutePages,
  pagesDir: string,
  defaultLocale: string
): RouteOptionsResolver {
  return (route, localeCodes): ComputedRouteOptions | null => {
    const options: ComputedRouteOptions = {
      locales: localeCodes,
      paths: {}
    }
    const pattern = new RegExp(`${pagesDir}/`, 'i')
    // TODO: we might be needed support for vite, this is for webpack
    const chunkName = route.chunkName ? route.chunkName.replace(pattern, '') : route.name
    const pageOptions = chunkName ? pages[chunkName] : undefined
    // Routing disabled
    if (pageOptions === false) {
      return null
    }
    // Skip if no page options defined
    if (!pageOptions) {
      return options
    }

    // Remove disabled locales from page options
    options.locales = options.locales.filter(locale => pageOptions[locale] !== false)

    // Construct paths object
    for (const locale of options.locales) {
      const customLocalePath = pageOptions[locale]
      if (typeof customLocalePath === 'string') {
        // Set custom path if any
        options.paths[locale] = customLocalePath
        continue
      }

      const customDefaultLocalePath = pageOptions[defaultLocale]
      if (typeof customDefaultLocalePath === 'string') {
        // Set default locale's custom path if any
        options.paths[locale] = customDefaultLocalePath
      }
    }

    return options
  }
}
