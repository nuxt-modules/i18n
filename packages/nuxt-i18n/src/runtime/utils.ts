import { isVue3 } from 'vue-demi'
import { useRequestHeaders } from '#app'
import { findBrowserLocale } from 'vue-i18n-routing'

import type { Composer } from '@intlify/vue-i18n-bridge'
import type { NuxtI18nInternalOptions } from '#build/i18n.options.mjs'

/**
 * Parses locales provided from browser through `accept-language` header.
 *
 * @param input - Accept-Language header value.
 * @return An array of locale codes. Priority determined by order in array.
 */
export function parseAcceptLanguage(input: string): string[] {
  // Example input: en-US,en;q=0.9,nb;q=0.8,no;q=0.7
  // Contains tags separated by comma.
  // Each tag consists of locale code (2-3 letter language code) and optionally country code
  // after dash. Tag can also contain score after semicolon, that is assumed to match order
  // so it's not explicitly used.
  return input.split(',').map(tag => tag.split(';')[0])
}

export async function loadAndSetLocale(newLocale: string, i18n: Composer /*, { initialSetup = false } = {}*/) {
  if (!newLocale) {
    return
  }

  // abort if different domains option enabled
  // if (!initialSetup && i18n.differentDomains) {
  //   return
  // }

  const oldLocale = i18n.locale.value as string
  if (newLocale === oldLocale) {
    return
  }

  i18n.locale.value = newLocale
  console.log('loadAndSetLocale', newLocale, oldLocale, i18n)
}

export function getBrowserLocale(options: Required<NuxtI18nInternalOptions>, context?: any): string | undefined {
  let ret: string | undefined

  if (process.client) {
    if (navigator.languages) {
      // get browser language either from navigator if running on client side, or from the headers
      ret = findBrowserLocale(options.__normalizedLocales, navigator.languages as string[])
    }
  } else if (process.server) {
    if (!isVue3) {
      if (context.req && typeof context.req.headers['accept-language'] !== 'undefined') {
        ret = findBrowserLocale(
          options.__normalizedLocales,
          parseAcceptLanguage(context.req.headers['accept-language'])
        )
      }
    } else {
      const headers = useRequestHeaders(['accept-language'])
      if (headers['accept-language']) {
        ret = findBrowserLocale(options.__normalizedLocales, parseAcceptLanguage(headers['accept-language']))
      }
    }
  }

  return ret
}
