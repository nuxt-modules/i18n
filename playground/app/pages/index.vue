<template>
  <div>
    <h1>{{ $t('welcome') }}</h1>
  </div>
</template>
