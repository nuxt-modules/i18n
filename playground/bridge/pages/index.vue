<script setup lang="ts">
import { useI18n } from 'vue-i18n-bridge'
import { localePath, switchLocalePath, useI18nHead } from 'vue-i18n-routing'

const { t, locale, locales, getBrowserLocale } = useI18n()
const i18nHead = useI18nHead({ addSeoAttributes: true })

console.log('getBrowserLocale', getBrowserLocale())

useMeta({
  title: computed(() => t('hello', { name: 'nuxt bridge' })),
  htmlAttrs: computed(() => ({
    lang: i18nHead.value.htmlAttrs!.lang
  })),
  link: computed(() => [...(i18nHead.value.link || [])]),
  meta: computed(() => [...(i18nHead.value.meta || [])])
})

function getLocaleName(code) {
  const locale = locales.value.find(i => i.code === code)
  return locale ? locale.name : code
}

const availableLocales = computed(() => {
  return locales.value.filter(i => i.code !== locale.value)
})
</script>

<template>
  <div>
    <h1>Demo: Nuxt Bridge</h1>
    <h2>{{ $t('hello', { name: 'nuxt bridge' }) }}</h2>
    <h2>Pages</h2>
    <nav>
      <NuxtLink :to="localePath('/')">Home</NuxtLink> |
      <NuxtLink :to="localePath('/about')">About</NuxtLink>
    </nav>
    <h2>Current Language: {{ getLocaleName(locale) }}</h2>
    <h2>Select Languages</h2>
    <nav>
      <span v-for="locale in availableLocales" :key="locale.code">
        <NuxtLink :to="switchLocalePath(locale.code) || ''">{{ locale.name }}</NuxtLink> |
      </span>
    </nav>
    <h2>I18n Head</h2>
    <div>
      <p>{{ i18nHead }}</p>
    </div>
  </div>
</template>
