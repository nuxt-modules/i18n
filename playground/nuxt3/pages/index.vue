<script setup lang="ts">
import { useI18n } from 'vue-i18n'
import { localePath, switchLocalePath, useI18nHead, useBrowserLocale } from '#i18n'

const { t, locale, locales } = useI18n()

console.log('useBrowserLocale', useBrowserLocale())

const i18nHead = useI18nHead({ addSeoAttributes: true })
useMeta({
  title: computed(() => t('hello', { name: 'nuxt bridge' })),
  htmlAttrs: computed(() => ({
    lang: i18nHead.value.htmlAttrs!.lang
  })),
  link: computed(() => [...(i18nHead.value.link || [])]),
  meta: computed(() => [...(i18nHead.value.meta || [])])
})

function getLocaleName(code) {
  const locale = locales.value.find(i => i.code === code)
  return locale ? locale.name : code
}

const availableLocales = computed(() => {
  return locales.value.filter(i => i.code !== locale.value)
})
</script>

<template>
  <div>
    <h1>Demo: Nuxt 3</h1>
    <h2>{{ $t('hello', { name: 'nuxt3' }) }}</h2>
    <h2>Pages</h2>
    <nav>
      <NuxtLink :to="localePath('/')">Home</NuxtLink> |
      <NuxtLink :to="localePath('/about')">About</NuxtLink>
    </nav>
    <h2>Current Language: {{ getLocaleName(locale) }}</h2>
    <h2>Select Languages</h2>
    <nav>
      <span v-for="locale in availableLocales" :key="locale.code">
        <NuxtLink :to="switchLocalePath(locale.code) || ''">{{ locale.name }}</NuxtLink> |
      </span>
    </nav>
    <h2>I18n Head</h2>
    <div>
      <p>{{ i18nHead }}</p>
    </div>
  </div>
</template>
