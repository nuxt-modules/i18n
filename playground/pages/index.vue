<script setup lang="ts">
import { computed } from 'vue'

// import { useLocalePath, useSwitchLocalePath, useLocaleHead, useBrowserLocale } from '#i18n'
import { LocaleObject } from '#i18n'

const route = useRoute()
const { t, strategy, locale, locales, localeProperties, setLocale, finalizePendingLocaleChange } = useI18n()
const localePath = useLocalePath()
const switchLocalePath = useSwitchLocalePath()
const getRouteBaseName = useRouteBaseName()

// route.meta.pageTransition.onBeforeEnter = async () => {
//   await finalizePendingLocaleChange()
// }

console.log('route base name', getRouteBaseName())
console.log('useBrowserLocale', useBrowserLocale())
console.log('localeProperties', localeProperties)
console.log('foo', t('foo'))
console.log('message if local layer merged:', t('layerText'))
console.log('message if github layer merged:', t('layer-test-key'))

function getLocaleName(code: string) {
  const locale = (locales.value as LocaleObject[]).find(i => i.code === code)
  return locale ? locale.name : code
}

const availableLocales = computed(() => {
  return (locales.value as LocaleObject[]).filter(i => i.code !== locale.value)
})

definePageMeta({
  title: 'pages.title.top',
  middleware: () => {
    const localePath2 = useLocalePath()
    console.log('middleware', localePath2({ name: 'blog' }))
  },
  pageTransition: {
    name: 'page',
    mode: 'out-in',
    onBeforeEnter: async () => {
      const { finalizePendingLocaleChange } = useNuxtApp().$i18n
      await finalizePendingLocaleChange()
      console.log('onBeforeEnter')
    }
  }
})
</script>

<template>
  <div>
    <h1>Demo: Nuxt 3</h1>
    <h2>{{ $t('hello', { name: 'nuxt3' }) }}</h2>
    <p>{{ $t('bar.buz', { name: 'buz' }) }}</p>
    <h2>Pages</h2>
    <nav>
      <NuxtLink :to="localePath('/')">Home</NuxtLink> | <NuxtLink :to="localePath({ name: 'about' })">About</NuxtLink> |
      <NuxtLink :to="localePath({ name: 'blog' })">Blog</NuxtLink> |
      <NuxtLink :to="localePath({ name: 'category-id', params: { id: 'foo' } })">Category</NuxtLink> |
      <NuxtLink :to="localePath({ name: 'history' })">History</NuxtLink>
    </nav>
    <h2>Current Language: {{ getLocaleName(locale) }}</h2>
    <h2>Current Strategy: {{ strategy }}</h2>
    <h2>Select Languages with switchLocalePath</h2>
    <nav>
      <span v-for="locale in availableLocales" :key="locale.code">
        <NuxtLink :to="switchLocalePath(locale.code) || ''">{{ locale.name }}</NuxtLink> |
      </span>
    </nav>
    <h2>Select Languages with setLocale</h2>
    <nav>
      <span v-for="locale in availableLocales" :key="locale.code">
        <a href="javascript:void(0)" @click="setLocale(locale.code)">{{ locale.name }}</a> |
      </span>
    </nav>
    <p>{{ $t('settings.profile') }}</p>
    <p>{{ $t('tag') }}</p>
  </div>
</template>

<style scoped>
.page-enter-active,
.page-leave-active {
  transition: opacity 1s;
}
.page-enter,
.page-leave-active {
  opacity: 0;
}
</style>
