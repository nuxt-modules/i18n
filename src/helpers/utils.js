/**
 * @typedef {import('../../types/internal').ResolvedOptions} ResolvedOptions
 */

/**
 * Retrieve page's options from the module's configuration for a given route
 *
 * @typedef {{ locales: readonly string[], paths: Record<string, string>} } ComputedPageOptions
 *
 * @param {import('@nuxt/types/config/router').NuxtRouteConfig} route
 * @param {ResolvedOptions['pages']} pages Pages options from module's configuration
 * @param {ResolvedOptions['localeCodes']} localeCodes
 * @param {string} pagesDir Pages dir from Nuxt's configuration
 * @param {ResolvedOptions['defaultLocale']} defaultLocale Default locale from Nuxt's configuration
 * @return {ComputedPageOptions | false} Page options
 */
export function getPageOptions (route, pages, localeCodes, pagesDir, defaultLocale) {
  /** @type {ComputedPageOptions} */
  const options = {
    locales: localeCodes,
    paths: {}
  }
  const pattern = new RegExp(`${pagesDir}/`, 'i')
  const chunkName = route.chunkName ? route.chunkName.replace(pattern, '') : route.name
  const pageOptions = chunkName ? pages[chunkName] : undefined
  // Routing disabled
  if (pageOptions === false) {
    return false
  }
  // Skip if no page options defined
  if (!pageOptions) {
    return options
  }

  // Remove disabled locales from page options
  options.locales = options.locales.filter(locale => pageOptions[locale] !== false)

  // Construct paths object
  for (const locale of options.locales) {
    const customLocalePath = pageOptions[locale]
    if (typeof customLocalePath === 'string') {
      // Set custom path if any
      options.paths[locale] = customLocalePath
      continue
    }

    const customDefaultLocalePath = pageOptions[defaultLocale]
    if (typeof customDefaultLocalePath === 'string') {
      // Set default locale's custom path if any
      options.paths[locale] = customDefaultLocalePath
    }
  }

  return options
}

/**
 * @param {string} routePath
 * @param {boolean | undefined} trailingSlash
 * @param {boolean} [isChildWithRelativePath] Indicates if it is a child route that has relative path
 * @return {string}
 */
export function adjustRouteDefinitionForTrailingSlash (routePath, trailingSlash, isChildWithRelativePath) {
  return routePath.replace(/\/+$/, '') + (trailingSlash ? '/' : '') || (isChildWithRelativePath ? '' : '/')
}
