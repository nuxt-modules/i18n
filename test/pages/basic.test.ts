import { vi, test, expect } from 'vitest'
import { localizeRoutes } from 'vue-i18n-routing'
import { getRouteOptionsResolver } from '../../src/pages'
import fs from 'node:fs'
import { getNuxtOptions } from './utils'

import type { NuxtI18nOptions } from '../../src/types'
import type { NuxtPage } from './utils'

test('localized', async () => {
  vi.spyOn(fs, 'readFileSync').mockReturnValue('')

  const options = getNuxtOptions({})
  const pages: NuxtPage[] = [
    {
      path: '/about',
      file: '/path/to/nuxt-app/pages/about.vue',
      children: [
        {
          name: 'about',
          path: '',
          file: '/path/to/nuxt-app/pages/about/index.vue',
          children: []
        }
      ]
    },
    {
      name: 'index',
      path: '/',
      file: '/path/to/nuxt-app/pages/index.vue',
      children: []
    }
  ]
  const localizedPages = localizeRoutes(pages, {
    ...options,
    includeUprefixedFallback: false,
    optionsResolver: getRouteOptionsResolver('pages', options as Required<NuxtI18nOptions>)
  })
  expect(localizedPages).toMatchSnapshot()
})
