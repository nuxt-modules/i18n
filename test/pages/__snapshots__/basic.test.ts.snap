// Vitest Snapshot v1

exports[`localized 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "file": "/path/to/nuxt-app/pages/about/index.vue",
        "name": "about___en",
        "path": "",
      },
    ],
    "file": "/path/to/nuxt-app/pages/about.vue",
    "path": "/about",
  },
  {
    "children": [
      {
        "children": [],
        "file": "/path/to/nuxt-app/pages/about/index.vue",
        "name": "about___ja",
        "path": "",
      },
    ],
    "file": "/path/to/nuxt-app/pages/about.vue",
    "path": "/ja/about",
  },
  {
    "children": [
      {
        "children": [],
        "file": "/path/to/nuxt-app/pages/about/index.vue",
        "name": "about___fr",
        "path": "",
      },
    ],
    "file": "/path/to/nuxt-app/pages/about.vue",
    "path": "/fr/about",
  },
  {
    "children": [],
    "file": "/path/to/nuxt-app/pages/index.vue",
    "name": "index___en",
    "path": "/",
  },
  {
    "children": [],
    "file": "/path/to/nuxt-app/pages/index.vue",
    "name": "index___ja",
    "path": "/ja",
  },
  {
    "children": [],
    "file": "/path/to/nuxt-app/pages/index.vue",
    "name": "index___fr",
    "path": "/fr",
  },
]
`;
